datumaro/components/format_detection.py

```diff
@@ -331,3 +350,79 @@ def apply_format_detector(
         context.fail(f"root path {dataset_root_path} must refer to a directory")
 
     return detector(context) or FormatDetectionConfidence.MEDIUM
+
+class RejectionReason(Enum):
+    unmet_requirements = auto()
+    insufficient_confidence = auto()
+
+class RejectionCallback(Protocol):
+    def __call__(self,
+        format_name: str, reason: RejectionReason, human_message: Sequence[str],
+    ) -> Any:
+        ...
+
+def detect_dataset_format(
+    detectors: Iterable[Tuple[str, FormatDetector]],
+    path: str,
+    *,
+    rejection_callback: Optional[RejectionCallback] = None,
+) -> Sequence[str]:
+    """
+    Determines which format(s) the dataset at the specified path belongs to.
+
+    `detectors` is a list of (name, detector) tuples. The function applies all
+    of the detectors to the path, and returns all names for which the
+    corresponding detector matched. If a detector returns matches with less
+    confidence than another detector, then the corresponding name is omitted.
+
+    For formats that are rejected, `rejection_callback`, if not `None`,
+    is called to report the reason why.
+    """
+
+    if not osp.exists(path):
+        raise FileNotFoundError(f"Path {path} doesn't exist")
+
+    def report_insufficient_confidence(
+        format_name: str,
+        format_with_more_confidence: str,
+    ):
+        if rejection_callback:
+            rejection_callback(
+                format_name, RejectionReason.insufficient_confidence, [
+                    f"Another format ({format_with_more_confidence}) "
+                        "was matched with more confidence",
+                ]
+            )
+
+    max_confidence = 0
+    matches = []
+
+    for format_name, detector in detectors:
+        log.debug("Checking '%s' format...", format_name)
+        try:
+            new_confidence = apply_format_detector(path, detector)
+        except FormatRequirementsUnmet as cf:
+            human_message = cf.generate_message()
```

Comment:
```suggestion
        except FormatRequirementsUnmet as e:
            human_message = e.generate_message()
```
- `cf` is not very clear acronym (check failure?), is suggest to use the regular naming
- If it is a regular exception, we can probably use `human_message = str(e)`

Reply:
> cf is not a very clear acronym (check failure?), I suggest to use the regular naming

Yeah, I don't even remember why I called it that. It's probably a reference to a name I replaced a long time ago. Renamed.

> If it is a regular exception, we can probably use human_message = str(e)

Okay, I did that. I also changed the type of `human_message` to `str`, since I couldn't find a compelling reason for it to be a list.